src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp

```diff
@@ -217,3 +271,29 @@ void GoalieFSM::moveToGoalLine(const Update &event)
         TbotsProto::BallCollisionType::AVOID,
         AutoChipOrKick{AutoChipOrKickMode::OFF, 0.0}));
 }
+
+bool GoalieFSM::retrieveDone(const Update &event)
+{
+    Point ball_position = event.common.world_ptr->ball().position();
+    Point retrieve_destination =
+        event.common.world_ptr->field().friendlyDefenseArea().centre();
+    return comparePoints(ball_position, retrieve_destination, 0.05);
+}
+
+void GoalieFSM::retrieveFromDeadZone(
+    const Update &event, boost::sml::back::process<DribbleFSM::Update> processEvent)
+{
+    Point ball_position = event.common.world_ptr->ball().position();
+    Vector final_dribble_orientation =
+        event.common.world_ptr->field().enemyGoalCenter() - ball_position;
```

Comment:
could you limit the final position so that we don't excessively dribble. I noticed a failure on my computer regarding this.